ENH: Add class to write dta format 117 files

bashtage opened the pull request:
Add export for dta 117 files which add support for long strings
Refactor StataWriter to simplify new writer

closes #16450

- [x] closes #16450
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

bashtage committed:
ENH: Add class to write dta format 117 files

Add export for dta 117 files which add support for long strings
Refactor StataWriter to simplify new writer

closes #16450

pep8speaks commented:
Hello @bashtage! Thanks for submitting the PR.

 - In the file [`pandas/io/stata.py`](https://github.com/pandas-dev/pandas/blob/9717de4da43a7392e8b7096b34e0209669491efe/pandas/io/stata.py), following are the PEP8 issues :

> [Line 2110:17](https://github.com/pandas-dev/pandas/blob/9717de4da43a7392e8b7096b34e0209669491efe/pandas/io/stata.py#L2110): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare except'

bashtage commented:
Is it intentional that `stata.py` is not covered?  Seems a bit strange.

jreback commented on pandas/core/frame.py:
```diff
@@ -1801,6 +1802,23 @@ def to_stata(self, fname, convert_dates=None, write_index=True,
+
+        convert_strl : list, optional
+            List of column names to convert to string columns to Stata StrL
+            format. Only available if version is 117.  Storign strings in the
```
Storing

jreback commented:
> Is it intentional that stata.py is not covered? Seems a bit strange.

what does this mean? the coverage that prints on the issue is odd (and has always been that way), but you can click thru and see the actual reports.

jreback commented on pandas/core/frame.py:
```diff
@@ -1801,6 +1802,23 @@ def to_stata(self, fname, convert_dates=None, write_index=True,
 
+        version : {114, 117}
+            dta version to use in the output file.  Version 114 can be used
+            read by Stata 10 and later.  Version 117 can be read by Stata 13
```
should we just support version 117 and forward? how old are these respective versions?

jreback replied:
how much simplfication do we get from only supporting new format?

bashtage committed:
DOC: Clean up doc strings

Fix typo
Enhance compliance of related docstrings usign validator

bashtage committed:
BUG: Fix bugs in stata

Fix incorrect skipping in strl writer
Fix incorrect byteorder when exporting bigendian
Fix incorrect byteorder parsing when importing bigendian
Improve test coverage for errors

jreback commented on pandas/io/stata.py:
```diff
@@ -2307,3 +2384,553 @@ def _null_terminate(self, s, as_string=False):
+        # byteorder
+        bio.write(self._tag(byteorder == ">" and "MSF" or "LSF", 'byteorder'))
+        # number of vars, 2 bytes
+        assert self.nvar < 2 ** 16
```
maybe add a blank line before comments, easier to read

jreback commented on pandas/tests/io/test_stata.py:
```diff
@@ -5,21 +5,22 @@
 
 import numpy as np
+import pytest
+from pandas._libs.tslib import NaT
```
import from pandas

jreback reviewed (changes requested):
minor comments. Might consiser renaming StataWriter to StataWriter114 (and essentially make it private). Otherwise lgtm.

bashtage commented:
> minor comments. Might consiser renaming StataWriter to StataWriter114 (and essentially make it private). Otherwise lgtm.

I think this would need a dep cycle since `StataWriter` has been public.  Can do this if you want.

bashtage committed:
CLN: Fix imports

Fix import location and add whitespace

TomAugspurger commented on pandas/core/frame.py:
```diff
@@ -1801,6 +1802,23 @@ def to_stata(self, fname, convert_dates=None, write_index=True,
 
             .. versionadded:: 0.19.0
 
+        version : {114, 117}
```
Hmm, should these be strings? Is this exposed anywhere in stats itself? Do they use integers? (when I see version number, I think string).

TomAugspurger replied:
Huh, https://www.stata.com/support/faqs/data-management/save-for-previous-version/ seems to suggest that stata uses integers? `version(13)`. OK then, let's follow that.

TomAugspurger commented on pandas/core/frame.py:
```diff
@@ -1832,12 +1856,23 @@ def to_stata(self, fname, convert_dates=None, write_index=True,
-        writer = StataWriter(fname, self, convert_dates=convert_dates,
+        kwargs = {}
+        if version not in (114, 117):
+            raise ValueError('Only formats 114 and 117 supported.')
```
Would be nice to include the user passed `version` in the error message.

TomAugspurger commented on pandas/tests/io/test_stata.py:
```diff
@@ -1355,3 +1413,63 @@ def test_date_parsing_ignores_format_details(self, column):
+                             columns=['long1' * 10, 'long', 1])
+        original.index.name = 'index'
+
+        with warnings.catch_warnings(record=True) as w:  # noqa
```
What warnings are you catching here?

TomAugspurger reviewed (approved):
Couple small comments. OK to do as a followup (I'm tagging the RC momentarily.)

TomAugspurger commented:
@jreback are you OK with merging this as is? I'll fix the git conflict on merge.

jreback commented:
yep this is fine
if we need follow ups after RC that is fine

TomAugspurger commented:
K. @bashtage I'm going to fix the conflict and then merge. A followup PR would be welcome.

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into bashtage-strl-support
